src/cli/commands/force/lightning/local/device/create.ts

```diff
@@ -0,0 +1,258 @@
+        this.deviceName = deviceName;
+        this.deviceType = deviceType;
+        return Promise.resolve();
+    }
+
+    // check whether device name is unique (i.e a virtual device with the same name doesn't exist already)
+    private async checkUniqueDevice(): Promise<string> {
+        const requirement = CommonUtils.castAsRequirement(this);
+        const isAndroid = CommandLineUtils.platformFlagIsAndroid(
+            Create.instance.platform
+        );
+        const deviceName = Create.instance.deviceName;
+
+        const deviceAlreadyExists = isAndroid
+            ? await AndroidSDKUtils.hasEmulator(deviceName)
+            : (await IOSUtils.getSimulator(deviceName)) != null;
+
+        return deviceAlreadyExists
+            ? Promise.reject(
+                  util.format(requirement.unfulfilledMessage, deviceName)
+              )
+            : Promise.resolve(
+                  util.format(requirement.fulfilledMessage, deviceName)
+              );
+    }
+
+    // check whether device type is valid (i.e. it matches one of the possible values of available devie types)
+    private async checkValidDeviceType(): Promise<string> {
+        const requirement = CommonUtils.castAsRequirement(this);
+        const isAndroid = CommandLineUtils.platformFlagIsAndroid(
+            Create.instance.platform
+        );
+        const deviceType = Create.instance.deviceType;
+
+        const supportedDevices = isAndroid
+            ? androidConfig.supportedDevices
+            : await IOSUtils.getSupportedDevices();
+
+        const match = supportedDevices.find((device) => device === deviceType);
+
+        return match !== undefined
+            ? Promise.resolve(
+                  util.format(requirement.fulfilledMessage, deviceType)
+              )
+            : Promise.reject(
+                  util.format(
+                      requirement.unfulfilledMessage,
+                      deviceType,
+                      supportedDevices.join(', ')
+                  )
+              );
+    }
+
+    private async executeDeviceCreate(): Promise<any> {
+        const isAndroid = CommandLineUtils.platformFlagIsAndroid(
+            Create.instance.platform
+        );
+        return isAndroid
+            ? this.executeAndroidDeviceCreate()
+            : this.executeIOSDeviceCreate();
+    }
+
+    private async executeAndroidDeviceCreate(): Promise<any> {
+        const preferredPack = await AndroidSDKUtils.findRequiredEmulatorImages();
+        const emuImage = preferredPack.platformEmulatorImage || 'default';
+        const androidApi = preferredPack.platformAPI;
+        const abi = preferredPack.abi;
+        let requestedPort = await AndroidSDKUtils.getNextAndroidAdbPort();
+        // need to incr by 2, one for console port and next for adb
+        requestedPort =
+            requestedPort < androidConfig.defaultAdbPort
+                ? androidConfig.defaultAdbPort
+                : requestedPort + 2;
```

Comment:
I know it's a small amount of duplication with [code in `AndroidLauncher`](https://github.com/forcedotcom/lwc-dev-mobile/blob/70cde5061fa64852a2ffae59a254aeb0165543ec/src/common/AndroidLauncher.ts#L33-L39), but given the somewhat esoteric business logic behind it, should we pull this out into its own method?

Reply:
@khawkins looks like `AndroidSDKUtils.getNextAndroidAdbPort()` is only called in 2 places and in both places we increment the port..... so perhaps we should just put this extra increment logic back into `AndroidSDKUtils.getNextAndroidAdbPort()` ?

Reply:
That seems reasonable, since as you say they're effectively connected. We can always re-separate if we need to down the line...but I doubt it. 😉

Reply:
Actually never mind... this was never needed here to begin with. we don't use `requestedPort` here at all so I'll just remove it.

Reply:
lol I didn't even notice that!